[1, 2단계 - 체스] 짱수(장혁수) 미션 제출합니다.

zangsu opened the pull request:
안녕하세요, 영이! 대망의 체스 미션 리뷰를 받게 된 짱수입니다!!
페어와 제가 고민했던 부분들을 description과 코드에서 공유드립니다!

### 모든 `public` 메서드의 테스트를 작성해야 하는가?

#### 배경 : 
이번 미션 역시 최대한 TDD 사이클을 준수하기 위해 노력했습니다. 다만, 이전 2개의 미션에서 TDD를 연습해 보면서 어느 정도 익숙해졌다고 생각하여 TDD 사이클을 크게 구성해 보았어요. 그러다 위의 고민을 마주쳤어요.
페어의 의견을 반영해 최대한 모든 메서드의 테스트를 작성해 주었지만, 혼자 코드를 작성했다면 저는 테스트를 작성하지 않았을 것 같은데요. 

아래는 제 생각을 정리해 보았어요. 이 부분에 대한 영이의 생각이 궁금합니다!

#### 문제 : 
큰 단위의 로직이 테스트를 통과하기 위해 더 작은 단위의 메서드가 새로 필요한 경우가 존재했습니다. 저희의 경우 다음과 같은 흐름이 존재했어요.

1. `Path`의 기능을 구현하던 중 "전체 경로가 모두 비어있는지 확인"을 해 주는 기능이 필요해짐
2. `Path`가 `Step.isEmpty()`라는 메서드를 호출해 해결해야 한다는 것을 깨닫고, `Step.isEmpty()`를 생성
3. `Step`이 `SquareState.isEmpty()` 라는 메서드를 호출해 해결해야 한다는 것을 깨닫고 `SquareState.isEmpty()` 생성 및 구현
4. `Step.isEmpty()` 구현
5. `Path.isAllEmpty()` 구현

#### 모든 메서드의 동작을 테스트 할 때의 장단점 : 

장점 : 
- 이후 해당 메서드를 사용하는 다른 메서드가 생기더라도 해당 메서드의 동작을 신뢰할 수 있다.

단점 : 
- 단순히 하위 메서드를 호출하는 정도의 (`SquareState.isEmpty()`) 메서드의 테스트코드는 얻을 수 있는 이점보다 낭비되는 리소스가 더 크다.

zangsu committed:
docs : 요구사항 작성

zangsu committed:
feat : 체스 칸을 나타내는 기능 구현

zangsu committed:
feat : 기물의 이동 방향을 나타낼 Move 구현

zangsu committed:
feat : 방향이 대각선인지, 수직선인지 확인할 수 있는 기능 구현

zangsu committed:
refactor : 제자리 경로가 생성되지 않도록 변경

zangsu committed:
test : 테스트 추가

zangsu committed:
refactor: 경로의 책임을 갖는 객체를 Move -> Path로 변경

zangsu committed:
style : README.md 수정

zangsu committed:
feat : 기물 클래스 생성

zangsu committed:
feat : King의 움직임 구현

zangsu committed:
feat : Queen의 움직임 구현

zangsu committed:
feat : Rook의 움직임 구현

zangsu committed:
feat : Knight의 움직임 구현

zangsu committed:
feat : Bishop의 움직임 구현

zangsu committed:
feat : Pawn의 움직임 구현

zangsu committed:
style : 컨벤션 적용

zangsu committed:
feat : 초기 위치 생성 기능 구현

zangsu committed:
feat : 보드 출력 기능 구현

zangsu committed:
feat : 입력 기능 구현

zangsu committed:
feat : 움직임 기능 통합 구현

zangsu committed:
refactor : 사용되지 않는 메서드 삭제

zangsu committed:
feat : 잘못된 입력에 대해 재입력 구현

zangsu committed:
refactor : 생성자 역할 최소화

zangsu committed:
feat: 경로의 길이 범위 검증 구현

zangsu committed:
refactor : 중복되는 생성 시점의 예외를 상수처리

zangsu committed:
refactor : 폰 생성시 기본으로 색깔을 지정

zangsu committed:
refactor : 방향들 생성 시 Location 사용하도록 변경

zangsu committed:
style : 가독성을 위한 변경

- 필드명 변경
- 변수명 변경
- 메서드명 변경
- 상수 추출
- 메서드 추출

zangsu committed:
refactor : Pawn들의 중복된 로직을 추상 클래스로 추출

zangsu committed:
style : README.md 취소선 삭제

zangsu commented on src/main/java/chess/domain/board/Path.java:
```diff
@@ -0,0 +1,102 @@
+
+    public boolean categoryNumOf(int categorySize) {
+        return steps.stream()
+                .map(Step::getDirection)
```
### getter의 사용을 지양한다.

이번 미션에선 다음과 같은 요구사항이 존재했습니다.
> 게터/세터/프로퍼티를 쓰지 않는다.

저와 페어는 이 때 getter를 사용하지 않아야 하는 이유로 다음의 이유들을 생각했어요.
1. getter를 사용해 값을 꺼내는 것은 객체의 캡슐화를 깨뜨리는 행위이다.
2. getter를 통해 값을 꺼낼 수 있다면, 특정 클래스의 책임을 외부에서 구현하기 쉬워진다. 

그런데, `Path`가 포함하고 있는 방향의 종류 개수를 확인하는 것은 `Path`의 책임이라고 생각했어요.
그리고, 이 책임을 구현하기 위해서라면 getter를 사용하는 것이 불가피하다고 생각했습니다.
때문에 이 부분에서 getter를 남겨두었습니다.

zangsu committed:
style : 불필요한 공백 제거

zangsu committed:
refactor : 가독성 개선

zangsu commented on src/main/java/chess/domain/location/Location.java:
```diff
@@ -0,0 +1,70 @@
+import chess.domain.board.Direction;
+import java.util.Objects;
+
+//TODO 캐싱
```
저와 페어 모두 `Location`에서 캐싱을 하는 것이 유리할 것이라고 생각했습니다.
이 때 저는 조금 구현이 복잡해지더라도 `Map<Row, Map<Column, Location>> CACHED_LOCATION` 을 사용하자는 의견을 제안했어요. 반대로 페어는 `Map` 내부에 또다른 `Map`이 저장되는 것이 가독성을 크게 저해한다고 생각했어요.

저와 페어의 의견이 합치하지 못해 TODO로 남겨두었으며, 1\~2단계 머지 전까지 해결되지 않는다면 3\~4단계 구현에서 따로 진행해 볼 것 같습니다.

zangsu commented on src/main/java/chess/domain/piece/Pawn.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    @Override
+    public boolean canMove(Path path) {
+        boolean checkMove = checkMovable(path);
+        moved = true;
+        return checkMove;
+    }
```
폰은 첫 번째 움직임에서만 특수하게 두 칸을 움직일 수 있습니다. 때문에 다른 메서드들과는 다르게 기물이 움직이면서 상태를 변경시켜 주어야 했어요.

저는 `canMove()`라는 메서드 명과 다르게 실제로 움직임이 이루어지는 (`moved`의 상태가 변경) 것이 마음에 들지 않았고, 이 부분을 `Piece.move()`라는 또 하나의 추상 메서드로 추출하자는 의견을 냈습니다.
그리고, 페어는 폰이라는 하나의 기물만을 위해서 `move()`라는 새로운 추상 메서드가 추가되는 것이 마음에 들지 않았고요.
영이는 이 부분에 대해 어떻게 생각하시나요?

choijy1705 replied:
moved의 상태가 변경되는건 그렇다쳐도 
canMove라는 실제 움직임이 아닌 곳에서 
움직였다고 변경하는건 나중에 심각한 오류를 낼 수 있을것 같네요!😭
예를들면 canMove 까지 수행했는데 이후 로직이 알수없는 오류로 move 메서드가 수행이 안되버리면 움직이지 않았는데 움직인걸로 기록될 수 있을것 같네요!

zangsu replied:
동의합니다!! 다른 모든 기물에서 동작이 존재하지 않더라도, Command와 Query를 분리하기 위해 (맞겠죠...? ㅎㅎ) `move()` 메서드를 추가해 주었습니다.

zangsu reviewed (commented):
고민들을 정리해 두었습니다!!
편하실 때 한번 참고 부탁드립니다 🙇

zangsu commented on src/main/java/chess/domain/piece/BlackPawn.java:
```diff
@@ -0,0 +1,14 @@
+
+import chess.domain.board.Path;
+
+public class BlackPawn extends Pawn {
```
DM으로 여쭤 봤던 "설계에 대한 의견차이를 좁히지 못해 선택한 3안이 `BlackPawn` / `WhitePawn` 이었습니다.
혹여나 도움이 될까 하여 저와 페어의 의견을 대략적인 의사코드와 함께 공유해드립니다.

### 상황

폰은 색깔에 따라 이동할 수 있는 방향이 다릅니다. 이 때 "검정색 폰"과 "흰색 폰"이 어느 방향으로 이동할 수 있는지를 판단하는 것에 대한 설계가 갈렸습니다.

### 1안 (저의 의견입니다.)

결국, 폰에서 전진 방향은 "색"에 따라 결정이 되는데요.
때문에 `Color` 에서 전진 방향 (`BLACK` -> `DOWN`, `WHITE` -> `UP`) 을 가지고 있으면서, `color.isForward(Path path)`의 식으로 폰의 전진 방향을 판단해야 한다고 생각했습니다.

대략 다음과 같은 코드가 예상이 됩니다. 
```java
class Pawn {
    public boolean canMove(Path path) {
        //== 전진 방향 이외의 로직 생략... ==//
        if(this.color.isBackward(path)) {
            return false;
        }
        //== 전진 방향 이외의 로직 생략... ==//
    }
}

enum Color {
    BLACK (Direction.DOWN), 
    WHITE (Direction.UP);

    private final Direction forward;

    public boolean isBackward(Path path) {
        return path.isDifferentDistance(forward);
    }
}
```

### 2안 (페어의 의견입니다.)

페어의 경우 `Color`가 방향 확인에 대한 책임을 가지는 것 자체가 가독성이 저하되며 책임이 모호해 지는 것이라 주장하였습니다.
때문에 `Path`에 `Color`, 또는 "각 팀의 전진 방향" 과 같은 정보를 넘겨주어야 한다고 하였습니다.

대략 다음과 같은 코드가 예상 됩니다. (꼭 Color를 파라미터로 전달하기 보단, Path에게 해당 책임을 위임한다 정도로 이해해 주시면 좋을 것 같습니다.)
```java
class Pawn {
    public boolean canMove(Path path) {
        //== 전진 방향 이외의 로직 생략... ==//
        if(path.isDifferentDistance(this.color)) { 
            return false;
        }
        //== 전진 방향 이외의 로직 생략... ==//
    }
}

class Path {
    public boolean isDifferentDistance(Color color) {
        if(color.isBlack()){
            return isUpside();
        }
        return isDownside();
    }
}
```

### 정리

많은 크루들이 `Color`가 방향 결정에 대한 책임을 가지는 것이 이상하다, 또는 직관적이지 못하다 라는 의견을 주었고, 저 역시 "많은 사람들이 보고 쉽게 이해할 수 있는 코드"를 작성하는 것 역시 중요하다는 생각을 합니다. 이런 관점이라면 책임 분리와의 트레이드 오프를 고려해 볼 수 있겠죠.

다만, "`Color`가 가질 책임이 아니다" 라는 의견에는 공감이 되지 않아요. 오히려 지금은 2안의 `Path.isDifferentDistance()` 에 작성했던 조건 분기가 `Color`의 책임을 `Path`이 뺏어 왔기 때문에 생기는 분기라고 생각을 하고요.

또한, `Path` 클래스가 `Color`를 파라미터로 받아 진행 방향을 확인하는 메서드를 구현해야 할 이유도 납득이 되지 않아요. `Path`가 팀의 정보를 알 필요는 없지만 코드를 동작시키기 위해 해당 메서드를 욱여넣은 기분이 듭니다.

혹시 영이도 `Color` 가 방향 결정에 대한 책임을 가지는 것이 부자연스럽다고 생각이 드신다면, 설명을 부탁드리고 싶습니다 ㅠㅠ...

choijy1705 replied:
`color.isBackward(path)` 는 확실히 조금 어색하네요...
근데 이건 언어적 차이지 책임의 관점은 아닌것 같네요
color.isUpsideColor() 등등 메서드 명을 바꿔보면 자연스러워지지 않을까요??

우선 전체적으로 메서드 명이 이해하기 어려운것 같습니다....😭
여기 뿐만 아니라 다른 곳도 한번 고민해보시면 좋을것 같아요!

zangsu replied:
전체적으로 메서드 네이밍을 수정해 보았습니다!

zangsu reviewed (commented):
DM 으로 말씀드린 부분 작성해 두었습니다.
(의사 코드와 각 의견에 대한 생각들이 포함되어 조금 길어진 것 같습니다... 너그럽게 봐주시면 감사하겠습니다)

choijy1705 commented on src/main/java/chess/GameController.java:
```diff
@@ -0,0 +1,57 @@
+package chess;
```
controller는 왜 별도의 package를 만들지 않고
application과 동일한 레벨에 두었나요?

zangsu replied:
다른 패키지들은 동일한 관심사를 갖는 여러 클래스들을 묶어두어 해당 클래스를 탐색, 관리하기 편하도록 패키지들을 만들었는데요. 이 때 `Controller`는 전체 흐름을 관리하고 `view`와 도메인 사이의 데이터 변환의 책임만을 가지기 때문에 `application`과 비슷한 책임을 가지고 있다고 생각했던 것 같아요. (application과 동일한 레벨에 둔 것이 의도한 것은 아니었고, 굳이 패키지를 분리할 생각을 하지 못했습니다.)

영이의 리뷰를 보고 고민을 해 보니 `Application`은 단지 프로그램의 엔트리포인트 역할만을 할 뿐이고, `Controller`도 특정 책임을 갖는 클래스라면 패키지를 분리해 주는 것이 관리에 더 용이할 수 있겠네요!
게다가 다른 패키지들이 전부 domain (model), view 로 구분이 되고 있으니 동일하게 controller 패키지가 존재하는 것이 더 자연스러울 것 같기도 하고요!
패키지를 분리하겠습니다!

choijy1705 commented on src/main/java/chess/GameController.java:
```diff
@@ -0,0 +1,57 @@
+
+public class GameController {
+    private static final InputView inputView = new InputView();
+    private static final OutputView outputView = new OutputView();
```
```suggestion
    private static final InputView INPUT_VIEW = new InputView();
    private static final OutputView OUTPUT_VIEW = new OutputView();
```
static 변수는 대문자 snake case를 사용해주세요

choijy1705 commented on src/main/java/chess/GameController.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    private void playTurn() {
+        while (true) {
```
while문도 사용을 최대한 하지말라고 가이드 드리는데요
while(true)는 정말 안좋은 코드인것 같습니다!
무한루프의 가능성이 너무 많을것 같아요 ㅠ

zangsu replied:
`while`의 사용 자체를 제거하기는 어려웠던 것 같아요...
`while(true)`를 제거하기 위해 게임의 진행 상태를 나타내는 enum을 도입해 보았습니다!

choijy1705 commented on src/main/java/chess/GameController.java:
```diff
@@ -0,0 +1,57 @@
+    private void playTurn() {
+        while (true) {
+            String command = inputView.readCommand();
+            if ("end".equalsIgnoreCase(command)) {
```
"end", "start"는 상수로 관리하는게 좋을것 같네요

zangsu replied:
커멘드와 관련된 정보는 유지보수의 편의성을 위해 클래스 최상단에서 상수로 관리되는 것이 좋을 것 같다고 생각이 들어 `InputView`, `Controller` 에서 해당 부분을 상수로 추출하였습니다!

choijy1705 commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    private void initialPawnSetting(Map<Location, Piece> board) {
+        for (Column value : Column.values()) {
+            board.put(new Location(value, Row.TWO), new WhitePawn());
+        }
+
+        for (Column value : Column.values()) {
+            board.put(new Location(value, Row.SEVEN), new BlackPawn());
+        }
```
반복문을 두번 돌 필요 없을것 같네요!

zangsu replied:
앗... 수정했습니다!!

choijy1705 commented on src/main/java/chess/domain/board/SquareState.java:
```diff
@@ -0,0 +1,13 @@
+package chess.domain.board;
+
+public enum SquareState {
```
SqaureState 는 어떤걸 뜻하나요?ㅎㅎ
positionState 같은 단어가 더 의미를 잘 전달할것 같네요!

zangsu replied:
이 부분은 저와 페어 모두 체스에서의 한 칸을 부르는 일반적인 이름이 있는지 의문이 들어 구글과 챗 지피티를 통해 검색을 해 보았던 부분입니다. 
> Chess is an abstract strategy game that involves no hidden information and no elements of chance. It is played on a chessboard with 64 squares arranged in an 8×8 grid.

다만, 저와 페어 역시 직관적으로 "한 칸의 상태" 라는 것이 와닿지는 않았던 것 같아요.
무언가 도메인 용어를 사용한다면 README에 작성을 해 두고, 그렇지 않다면 조금 더 직관적인 이름을 사용하는 것이 좋을 수 있겠네요...

우선, 의미를 더 잘 전달할 수 있는 이름으로 변경해 보겠습니다!

choijy1705 commented on src/main/java/chess/domain/board/Path.java:
```diff
@@ -0,0 +1,102 @@
+        return steps.size() == size;
+    }
+
+    public boolean categoryNumOf(int categorySize) {
```
메서드 명이 잘 이해가 안가네요
categoryNum이면 적어도 int가 반환될것 같아요

choijy1705 commented on src/main/java/chess/domain/board/Path.java:
```diff
@@ -0,0 +1,102 @@
+                .anyMatch(Step::isOrthogonal);
+    }
+
+    public boolean hasPiecePathExcludedTarget() {
+        return !isEmptyPathExcludedTarget();
+    }
+
+    private boolean isEmptyPathExcludedTarget() {
+        return createPathExcludedTarget()
+                .stream()
+                .allMatch(Step::isEmpty);
+    }
```
정반대 의미의 메서드를 굳이 두개로 만들필요가 있을까요?

zangsu replied:
어라, 두 메서드 모두 외부에서 사용되는 줄 알고 가독성을 위해 `!` 사용 대신 메서드 명으로 조건을 나타내려고 했는데요...
`isEmptyPath...` 메서드가 바로 위에서밖에 사용되지 않고 있었네요 😕 

해당 부분 리팩토링 진행하였습니다!

choijy1705 commented on src/main/java/chess/domain/location/Column.java:
```diff
@@ -0,0 +1,65 @@
+    G(7),
+    H(8);
+
+    private static final Column[] columns = new Column[]{A, B, C, D, E, F, G, H};
```
배열은 사용하지 않는게 요구사항으로 알고 있어요

zangsu replied:
우선 해당 변수를 선언한 이유는 enum의 선언 순서에 의존적인 코드를 작성하는 것을 피하기 위함이었습니다. (enum의 선언 순서에 의존적인 코드가 좋지 않은 코드인가? 에 대한 기준은 아직 고민중이지만요...ㅎㅎ)
`values()`를 사용하는 것 만으론 순서가 중요하는 것을 나타내기 힘들 것 같아서 필드로 한번 더 순서를 명시해 의도를 나타내 보고 싶었습니다. 
대신 `columns`가 `values()`와 거의 비슷한 역할을 하는 만큼, `values()`의 반환값과 동일한 배열을 사용했던 것 같습니다. 

다만, 이후 해당 부분이 어떻게 더 사용될 지 모르는 단계이기에 확장성을 고려해 준다면 컬렉션을 사용하는 것이 더 좋은 선택이 될 수 있을 것 같네요. 
리팩토링을 진행해 보겠습니다.

choijy1705 commented on src/main/java/chess/domain/location/Column.java:
```diff
@@ -0,0 +1,65 @@
+        this.index = index;
+    }
+
+    public static Column of(String input) {
+        return Arrays.stream(values())
+                .filter(column -> column.isName(input))
+                .findAny()
+                .orElseThrow(() -> new IllegalArgumentException("잘못된 Column 입력입니다."));
+    }
```
배열이 아닌 Map으로 상수를 미리 만들어두면 o(1) 로 반환받을 수 있을것 같네요

zangsu replied:
정말 미묘한 차이지만, 의문이 생겨서 질문을 남겨봅니다!
지금은 입력 값을 enum 상수의 이름과 비교해서 적절한 `Column`을 가져오고 있는데요. 이 부분은 `view`에서 `Column`이 가지는 특성을 사용하는 것이라고 느껴집니다. 즉, 입력값 (view)이 도메인에 의존을 하고 있는 것 같아요.

그런데, 성능을 개선시키기 위해서 `Map<String, Column>`을 만들어 두는 순간, key로 사용될 `String`은 "입력에서 사용될 값"이라고 느껴지는 것 같아요. 더 정확히는 key로 사용될 값이 view에 의존하고 있다기 보단, key로 `String`을 가지는 `Map`을 만드는 것 자체가 view에 의존하는 것으로 느껴져요.

물론, 영이가 말해주신 대로 상수를 미리 만들어 두면 성능의 개선과 더불어 코드도 훨씬 간단해 질 것 같아요. 
아래와 같은 코드를 작성할 수 있겠네요!
```java
return Optional.ofNullable(COLUMNS.get(input))
        .orElseThrow(() -> new IllegalArgumentException("잘못된 Column 입력입니다."));
```

`Map<String, Column>`을 도메인이 가지도록 하는 것은 위와 같은 이유들로 충분히 합리적인 트레이드오프가 되는 걸까요? view를 위해 생겨난 코드라고 느껴지는 것이 과한 걱정인걸까요?

choijy1705 replied:
List보다는 Map이 나을것 같아요
다만 여기서는 근본적인 질문을 해봐야할것 같습니다!
enum에는 ordinal이 내장되어 있어요 하지만 일반적으로 사용하길 권장하지 않습니다
지금 인위적으로 List를 만드는 것도 같은 이유에서 권장되는 방법은 아닐것 같아요
한번 고민해보시면 좋을것 같아요!

zangsu replied:
아시겠지만, 제가 인위적으로 `List`를 생성한 것은 "순서가 유의미하다"는 것을 드러내기 위함 + enum 의 선언 순서에 의존하지 않고, `List`에서 관리되는 순서에 의존하기 위함이었습니다. `ordinal()`을 사용하지 않은 것 역시 비슷한 이유였어요.

솔직히 말하면 순서가 존재하는 것이 자연스러운 도메인에서 순서에 의존하는 것이 왜 문제인지는 아직까지 이해가 잘 안되기는 합니다...ㅎㅎ 물론 데이터의 변경이 존재하는 상황이라면 유지보수를 위해 다른 방법을 고려해 볼 수 있겠지만, 체스판이라는 도메인은 변경이 일어날 확률이 극히 드물기 때문에 클래스에 변경이 예상되지 않기도 하고요... 
java.time의 `DayOfWeek`, `Month`와 같은 클래스도 비슷하게 필연적으로 순서가 존재하는 도메인이기에 `values()`에 의존적인 코드가 존재한다고 생각해요.
이 부분에 대해서 너무 안일하게 생각하고 있는 것일까 걱정도 드네요.

다만, 이미 제 코드를 납득시키기 위해 벌써 영이와 두 번의 핑퐁이 존재하는 것을 고려한다면, 코드를 읽을 때 인지적 부담이 존재한다는 것은 부정할 수 없을 것 같다는 생각이 들었어요.

코딩은 문학이라는 말을 들어본 적이 있어요. 동일한 코드를 읽고 모두가 만족할 수는 없겠지만, 적어도 7~8명 정도가 만족한다면 좋은 코드일 것이라는 메시지였는데요. 저의 코드를 제 주변 크루들에게 설명하기 위해서도 비슷한 과정이 존재했다는 것은 적어도, 제 코드에 개선할 부분이 있다는 뜻이겠죠..ㅎㅎ

---

최대한 영이가 말해 주신 방법대로 해당 부분 리팩토링 해 보았습니다. 리팩토링 방법을 고민하던 중 굳이 순서에 의존하지 않아도 충분히 코드를 작성할 수 있다는 사실을 깨닫기도 했고요! 리팩토링 과정에서 영이가 머릿속에서 그리던 방향이 이런 방향이었을 것 같다는 생각이 들었어요. 부디 마음에 들었으면 좋겠어요!

choijy1705 commented on src/main/java/chess/domain/location/Row.java:
```diff
@@ -0,0 +1,66 @@
+    SEVEN(7),
+    EIGHT(8);
+
+    private static final Row[] ROWS = new Row[]
```
여기도 배열 사용 하고 있네요

zangsu replied:
(위의 코멘트와 동일합니다)

choijy1705 commented on src/main/java/chess/domain/piece/Pawn.java:
```diff
@@ -0,0 +1,44 @@
+public abstract class Pawn extends Piece {
+    abstract boolean isBackward(Path path);
+
+    private boolean moved = false;
```
생성자에서 초기화 해주세요

zangsu replied:
영이가 "질문"이 아닌 "요구"를 주셨던 것은 초기값의 선언 위치에 따라 달라지는 무언가 있을 것이라는 생각에 조금 더 찾아보게 되었습니다.

아래는 제가 찾아본 각 방법의 장점인데요.

- 클래스 바디에서 선언과 함께 초기화
	- 여러 생성자가 제공되더라도, 항상 해당 값이 동일하게 초기화 되어야 한다면 코드의 라인을 줄여 클래스를 가볍게 가져갈 수 있다.
- 생성자에서 변수를 초기화 한다면
	- 객체 생성 방법이 궁금할 때 생성자만 확인하면 되기 때문에 "생성자의 응집도"를 높여줄 수 있다.
	- 생성자에서 또다시 값이 초기화가 된다면 선언시에 사용한 값을 덮어쓰게 되므로 생성자에서 사용할 값을 명시해 두는 것이 사이드이펙트를 더 줄일 수 있다.

이렇게 정리해 두니 생성자에서 변수를 초기화 하는 것이 충분히 "클래스의 구현 코드 라인을 줄일 수 있다"는 장점과 트레이드오프 할 수 있을 만큼 매력적인 것 같아요. 초기화의 위치를 생성자로 옮기겠습니다!

혹시, 영이가 리뷰를 주신 근거 중에 제가 놓친 부분이 있을까요?

choijy1705 replied:
지금은 대부분의 도메인의 필드 개수가 적지만 
실무에서는 10개 이상의 필드를 가지고 있는 경우가 많습니다
이때 어느곳에서는 생성자에서 초기화하고 어떤곳은 필드에서 초기화 되버리면
필드의 값들을 캐치하지 못하는 경우가 생길 수 있을것 같아요!

찾아보신 방법도 충분한 이유가 될 수 있을것 같네요!👍

choijy1705 commented on src/main/java/chess/domain/piece/Queen.java:
```diff
@@ -0,0 +1,18 @@
+        if (path.hasPiecePathExcludedTarget()) {
+            return false;
+        }
+        return path.hasNoAllyAtTarget() && path.categoryNumOf(1);
```
caregoryNumOf는 아무리 봐도 무슨 의미인지 모르겠네요 ㅠㅠ

zangsu replied:
이번 미션이 유독 네이밍이 어려웠어요...ㅠㅠ
네이밍을 조금 더 개선해 봤는데, 효과가 있었으면 좋겠습니다 ㅠㅠ

choijy1705 commented on src/main/java/chess/view/OutputView.java:
```diff
@@ -0,0 +1,74 @@
+        return s.toLowerCase();
+    }
+
+    private String getPieceString(Piece piece) {
```
piece가 name을 필드로 가지고 있으면 안되나요?

zangsu replied:
해당 방법도 고민이 되었고, 결론을 내린 과정을 블랙잭 미션과 비교해 말씀드릴 수 있을 것 같은데요.

블랙잭 미션에서 카드를 출력하기 위해 사용하는 `A`, `2`, `K` 등은 이미 대부분의 카드에서 동일하게 표기하고 있기 때문에 도메인이 뷰의 책임을 가지고 있는 것 보단, 뷰에서 도메인의 특성값을 사용한다고 보는 것이 더 자연스럽다고 생각했어요. (물론 저는 view 에서 처리하긴 했지만요...ㅎㅎ)

아마, 영이는 `Piece`가 출력에서 사용할 `P`, `N`, `Q` 등도 위와 동일하게 해 보는 걸 추천하신 것 같은데요.
다만, 체스의 경우 각 기물을 위와 같이 표기하는 것이 일반적인가? 라고 생각했을 때 그렇지 않다고 결론을 내린 것 같습니다.
일반적으로 체스 기물을 표기할 때 범용적으로 사용되는 기호들이 존재하기에 위의 값들은 기물의 특성보단 출력 형식에 가깝다고 생각한 것 같아요.

그런데, 기물의 이름이 `King`, `Pawn` 이라는 것이 변경되지 않는다면 영이가 말했던 것 처럼 `name` 역시 기물의 특성이라 생각해야 하나 고민이 되긴 하네요...

choijy1705 replied:
도메인에서 name을 가지는건 전혀 어색하지 않을것 같네요
instanceof 이렇게 사용하는건 요구사항도 여러가지를 위반하는 것 같고
전체적으로 바람직하지않은 코드인것 같습니다ㅠㅠ😭

zangsu replied:
`instanceof`를 사용하는 대신 각 기물이 가진 데이터를 활용하라는 뜻이 맞았겠죠?
`String`보단 enum으로 타입을 관리하는 것이 다양한 측면에서 훨씬 유리할 것이라 생각해 `enum PieceType`을 생성해 보았습니다!

choijy1705 commented on src/main/java/chess/domain/board/Path.java:
```diff
@@ -0,0 +1,102 @@
+                .allMatch((Step::isUpside));
+    }
+
+    public boolean isTargetHasEnemy() {
```
hasTargetEnemy?

zangsu replied:
`isXxxAtXxx()` 로 변경해 보았습니다 :D

choijy1705 reviewed (changes requested):
안녕하세요 짱수!
리뷰의 양이 많아져서 한번 끊고 가려고합니다
메서드 명, 클래스 명이 쉽게 이해가지 않는 부분들이 있었는데
이부분 고민해서 조금 더 이해하기 쉽도록 네이밍 해주시면 좋을것 같아요!🙏

choijy1705 commented:
추가로 모든 메서드를 테스트 할 필요는 없습니다.
모든 메서드를 테스트 하는것이 절대로 테스트를 잘 했다고 볼 수 없을것 같아요!
필요하다고 생각하는 테스트만 진행해주셔도 될 것 같아요👏

zangsu committed:
refactor : static 변수의 네이밍을 대문자로 변경

zangsu committed:
refactor : 입력 커멘드 관련 정보를 상수 추출

zangsu committed:
refactor : Controller  패키지 분리

zangsu committed:
refactor : 변수 초기화를 클래스 바디 -> 생성자로 변경

zangsu committed:
refactor(Path) : 사용되지 않는 메서드 삭제

zangsu committed:
refactor(Column, Row) : 캐싱 리스트를 배열 -> List 로 변경

zangsu committed:
refactor : 가독성 개선을 위한 네이밍 변경

zangsu committed:
refactor : 움직임 확인과 움직임 로직 분리

zangsu committed:
refactor : 불필요한 반복문 분리 제거

zangsu committed:
refactor : Controller 로직 개선

- while(true) 제거
- 게임 진행을 위한 게임 상태, 커맨드 객체 생성

choijy1705 commented on src/main/java/chess/controller/GameController.java:
```diff
@@ -0,0 +1,55 @@
+    private static final InputView INPUT_VIEW = new InputView();
+    private static final OutputView OUTPUT_VIEW = new OutputView();
+
+    private Board board = new Board();
+    private GameState state = GameState.PLAYING;
```
생성자에서 초기화해주세요!

zangsu replied:
이번의 경우에는 생성자가 존재하지 않던 클래스라서 고민 하던 부분인데요.
생성자 구현 후 생성자에서 초기화 하도록 변경하였습니다!

choijy1705 commented on src/main/java/chess/controller/GameController.java:
```diff
@@ -0,0 +1,55 @@
+        try {
+            play();
+        } catch (RuntimeException exception) {
+            OUTPUT_VIEW.printException(exception);
```
의도하지 않은 에러일 경우
유저에게 보여지는 메세지가 알기 어렵게 전달 될 수 있을것 같네요

zangsu replied:
커스텀 예외를 염두에 두고 말씀하신건가요, 아니면 `RuntimeException` 말고 더 세분화된 메시지를 말씀하신건가요?

전자의 경우라면, 프로젝트 규모에 비해 구현해야 할 커스텀 예외의 종류가 너무 많아진다고 느껴 커스텀 예외를 사용하는 대신 `IllegalArgumentsException / IllegalStateException` 두 종류만 사용하는 대신 메시지로 구분하려고 하였습니다. 

후자의 경우로 이해 후 사용중인 위의 두 종류 예외만 처리하도록 변경하였습니다.

zangsu replied:
+) 리팩토링을 진행하며 `NumberFormatException`이 발생했을 때 제가 지정해 주지 않은 예외 메시지가 발생되는 점을 확인하였습니다. `try-catch`를 사용해 새로 예외를 생성해 던져주었습니다.

choijy1705 replied:
NumberFormatException 처럼 지정해주지 않은 에러는 많은 경우에 날 수 있습니다. 그러면 지금 구조에서는 어떤 에러가 원인지 유저입장에서는 잘 볼 수없는 raw한 에러가 내려갈것 같네요!

controller단에서는 runtimeException 혹은 Exception을 catch하여서 적어도 유저가 보기 편한 메세지가 전달될 수 있도록 할 필요가 있을것 같아요🤔

zangsu replied:
사용자 편의를 위해 제가 의도하지 않은 예외에 대해서도 적당한 예외 메시지를 던져주는 것이 좋겠다는 말씀이시죠?  

유저에게 에러의 원인을 전달하기는 어렵겠지만, 적어도 사용자 친화적인 메시지를 전달해 줄 수는 있을 것 같아요. 이 방향으로 반영해 보겠습니다!

choijy1705 commented on src/main/java/chess/controller/GameController.java:
```diff
@@ -0,0 +1,55 @@
+            play();
+        } catch (RuntimeException exception) {
+            OUTPUT_VIEW.printException(exception);
+            run();
```
play()?

choijy1705 replied:
잘못입력하면 계속해서 체스게임을 시작합니다가 떠서
게임 진행이 어렵네요 ㅠㅠ

zangsu replied:
출력 부분 분리를 위해 나뉘어져 있던 메서드를 과하게 나뉘어진 메서드로 착각해서 리팩토링 했었네요...ㅠ
`play`를 호출한다면 두번째 잘못된 입력부터는 처리되지 않고 종료되었을 거에요..ㅠㅠ

다시 해당 부분 분리해서 개선하였습니다!

choijy1705 commented on src/main/java/chess/controller/GameController.java:
```diff
@@ -0,0 +1,55 @@
+        if (command == Command.END) {
+            state = GameState.END;
+        }
+        if (command == Command.START) {
+            initBoard();
+        }
```
처음에만 start입력시 초기화를 하고
진행중인 게임이 있다면 다른 입력을 받도록 처리하면 좋을것 같네요!

zangsu replied:
각 상태들이 한정적인 기능만을 제공하도록 완전히 분리해 주는 것이 좋을 것 같아 다음과 같이 리팩토링 하였습니다. 

- start
  - 게임 시작
  - 진행중인 게임이 있을 경우 시작 불가
- move
  - 움직임 제공
  - 게임이 진행중일 경우에만 가능
- end
  - 게임 종료
  - 모든 경우에 게임을 종료시키고 프로그램을 종료함

choijy1705 commented on src/main/java/chess/domain/location/Location.java:
```diff
@@ -0,0 +1,70 @@
+            Row row = Row.of(input.substring(1, LOCATION_INPUT_LENGTH));
+            return new Location(column, row);
+        } catch (IllegalArgumentException exception) {
+            throw WRONG_LOCATION_INPUT_EXCEPTION;
```
에러를 상수로 관리하면 
나중에 디버깅할때 찾기 번거로울것 같습니다!

choijy1705 replied:
그리고 여기서 try catch를 할필요가 있을까요??

zangsu replied:
> 에러를 상수로 관리하면
나중에 디버깅할때 찾기 번거로울것 같습니다

잘못된 위치 입력에 대한 에러가 중복되어 분포해 있길래 상수로 추출해 보았는데요... 디버깅에 어렵겠군요!
다시 명시적으로 예외를 발생시키겠습니다.

> 그리고 여기서 try catch를 할필요가 있을까요??

이 부분은 각각 잘못된 `Column` 입력과 잘못된 `Row` 입력을 더 큰 범위의  잘못된 `Location` 입력으로 처리해주려던 부분인데요. 생각해 보니 이 부분 역시 필요 없을 것 같아 제거해 주었습니다.

choijy1705 commented on src/main/java/chess/domain/location/Location.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private static void validateInput(String input) {
+        if (input == null || input.isBlank()) {
+            throw WRONG_LOCATION_INPUT_EXCEPTION;
+        }
+        if (input.length() != LOCATION_INPUT_LENGTH) {
+            throw WRONG_LOCATION_INPUT_EXCEPTION;
+        }
```
공통적인 에러라서 하나로 묶어도 될것 같네요!

zangsu replied:
에러가 중복되었지만, 리팩토링을 위해 다시 들여다 보니 "빈값 입력" / "잘못된 값 입력"으로 구분이 되고 있었어요. 그래서 하나로 묶는 대신, 서로 다른 메시지의 예외를 발생시키도록 변경하였습니다.

choijy1705 commented on src/main/java/chess/domain/location/Location.java:
```diff
@@ -0,0 +1,70 @@
+        try {
+            validateInput(input);
+            Column column = Column.of(input.substring(0, 1));
+            Row row = Row.of(input.substring(1, LOCATION_INPUT_LENGTH));
```
2는 LOCATION_INPUT_LENGTH로 상수처리하고 1은 왜 상수처리 하지 않았는지 궁금하네요🤔

zangsu replied:
아아....! 제 기억으로는 이 부분에서 페어와 의견이 갈렸던 것으로 기억해요...
위 코드는 제 의견이었고, 아래 코드는 페어 의견이었는데 비교를 위해 한 곳에서만 상수를 추출해 두고는 그대로 넘어가버렸네요 ㅠㅠㅠ

저는 이 부분이 "의미를 가지는 상수" 이기는 하지만, 상수로 추출해 네이밍을 해 주는 것 보다 매직 넘버 자체로 읽히는 것이 훨씬 가독성이 좋다고 생각해 상수를 제거하고 매직넘버로 바꾸어 주었습니다.

choijy1705 commented on src/main/java/chess/domain/location/Row.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    private boolean isRank(String input) {
+        return rank == Integer.parseInt(input);
```
input에 숫자가 아닌 다른 값이 들어갔을때는 
어떠한 에러 처리도 되어있지 않을것 같네요!
애초에 of에서 int로 rank를 받는다면 
previous, next로직과 함께 전반적으로 개선되지 않을까 싶네요!

zangsu replied:
관련 변경사항입니다. 
- `Column`, `Row`의 정적 팩토리 메서드 이름을 `of` 에서 `findByXXX`로 변경하였습니다.
- `Row`의 정적 팩토리 메서드를 오버로딩하여 `int`를 받을 수 있도록 하였습니다. 
  - 두 메서드가 모두 `public`인 이유는 객체 생성에 있어서는 생성자가 많을 수록 유연한 객체 활용이 가능하다고 생각하기 때문입니다.

choijy1705 commented on src/main/java/chess/controller/GameController.java:
```diff
@@ -0,0 +1,55 @@
+        }
+    }
+
+    private void playTurn(Command command) {
```
start를 하지 않아도 move를 할 수 있고 게임을 종료시킬 수 있네요😭

zangsu replied:
`Command`를 도입하던 시점에서 "사용자 편의를 위해서라면 "start" 명령어를 체스 판을 초기화 하고, 게임을 재시작하기 위해서 사용하고 첫 입력부터 이동/종료를 가능하게 하면 어떨까? 라는 생각을 했어요.
다시 생각해 보니 게임의 진행 상태를 명확하게 나누고 수행 가능한 동작에 제약을 걸어두는 것이 사이드 이펙트를 예방할 수 있는 것 같이 느껴지기도 하네요.

---

게임을 시작해야만 움직임을 제공할 수 있도록 다시 변경하였습니다.
다만, 게임 종료의 경우 게임을 시작하지 않더라도 프로그램 종료를 위해 사용 가능하도록 열어두었습니다!

choijy1705 reviewed (changes requested):
짱수 게임의 진행 상태를 나타내도록 enum을 도입해주셨는데요!
start, move, end에서 여러가지 치명적인 허점들이 많이 보였습니다😭
리뷰를 바탕으로 한번 점검해보시면 좋을것 같습니다
요거 반영후 piece들의 움직임들을 리뷰드리겠습니다!

zangsu committed:
refactor : 예외 처리 개선

- 예외 상수처리 제거
- 불필요한 예외 처리 (`Row`, `Column`의 예외를 `Location`의 예외로 변경) 제거
- 예외 종류 구분 (빈값 입력 / 잘못된 입력)

zangsu committed:
refactor : 불필요한 상수 제거 및 매직넘버 사용

zangsu committed:
refactor : 게임 진행에 대한 정보를 관리할 `ChessGame` 구현

zangsu committed:
refactor : 행, 열 생성시 순서에 의존하지 않도록 변경

zangsu committed:
refactor : 각 기물이 기물 종류를 저장하도록 변경

zangsu committed:
style : 컨벤션 적용

choijy1705 commented on src/main/java/chess/controller/ChessGame.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.Map;
+
+public class ChessGame {
+    private Board board = new Board();
+    private GameState state = GameState.PREPARING;
```
생성자에서의 초기화가 아직 습관이 안되어있는것 같네요!

zangsu replied:
앗... 세 번의 리뷰 중 세 번 같은 리뷰를 받았네요...ㅠㅠ
이번에 새롭게 리뷰받은 부분이라 습관이 안되어 있기도 했지만, 좀 더 신경을 써 보겠습니다...!

choijy1705 commented on src/main/java/chess/controller/ChessGame.java:
```diff
@@ -0,0 +1,38 @@
+    public void initBoard() {
+        if (state == GameState.PLAYING) {
+            throw new IllegalStateException("이미 게임이 진행중입니다. 새로운 게임을 시작하려면 게임을 종료 후 재시작 해주세요");
+        }
```
기획적인 부분이지만
게임을 새롭게 진행하고 싶어서 start를 입력할수 있지 않을까요??
다시 종료시키고 새로 시작하는 것은 유저 입장에서는 조금 번거로울 것 같아요🤔

zangsu replied:
> 처음에만 start입력시 초기화를 하고 진행중인 게임이 있다면 다른 입력을 받도록 처리하면 좋을것 같네요!

이 부분을 제가 이해를 잘못했던 것 같네요..! 영이의 의도는 "정말 재시작 하겠습니까?" 정도의 입력을 추가해 주는 것이었을까요?
게임 초기화 기능을 다시 제공해 보겠습니다!

choijy1705 commented on src/main/java/chess/controller/ChessGame.java:
```diff
@@ -0,0 +1,38 @@
+        state = GameState.PLAYING;
+    }
+
+    public void tryMove(Location source, Location target) {
```
try라는 prefix로 실패할 수 있다는 가능성을 굳이 메서드 명으로 나타낼 필요는 없을것 같네요!

zangsu replied:
아하, 어짜피 이동할 수 없는 경우 예외를 던져주니 `move()`라는 메서드 명이 훨씬 직관적이겠네요. 
반면에 `tryMove()`라는 메서드명은 `boolean`으로 이동 여부를 던져줄 것 같은 뉘앙스이고요.

피드백 반영하였습니다!

choijy1705 commented on src/main/java/chess/controller/GameController.java:
```diff
@@ -0,0 +1,66 @@
+        Location source = Location.of(sourceInput);
+        Location target = Location.of(targetInput);
+        chessGame.tryMove(source, target);
+        OUTPUT_VIEW.printBoard(chessGame.getBoard());
```
적절히 개행 해주시면 더 읽기 좋을것 같아요!

choijy1705 commented on src/main/java/chess/domain/location/Column.java:
```diff
@@ -0,0 +1,70 @@
+    G(7),
+    H(8);
+
+    private static final Map<String, Column> CACHED_COLUMNS_BY_NAME = new HashMap<>();
+
+    static {
+        for (Column column : values()) {
+            CACHED_COLUMNS_BY_NAME.put(column.name(), column);
+        }
+    }
```
```suggestion
        private static final Map<String, Column> CACHED_COLUMNS_BY_NAME = Arrays.stream(values())
            .collect(Collectors.toMap(Enum::name, Function.identity()));
```
좀 더 간단하게 표현할 수 있을것 같아요🙏

zangsu replied:
와, 지금까지 `toMap()`을 사용하면 더 가독성이 떨어지던 경험이 많아서 자연스럽게 제외시켰던 것 같네요...!

바로 반영하겠습니다!

choijy1705 commented on src/main/java/chess/domain/location/Location.java:
```diff
@@ -0,0 +1,63 @@
+        this.row = row;
+    }
+
+    public Location move(Direction direction) {
+        Column movedColumn = this.column.move(direction);
+        Row movedRow = this.row.move(direction);
+        return new Location(movedColumn, movedRow);
+    }
+
```
어떤 곳에서는 불변객체를 위해 객체를 반환하고 
어떤 곳에서는 그냥 객체의 값만을 변경하고 있는것 같아요 (예를들면 chessGame 의 endGame)
기조를 전체적으로 통일하면 좋을것 같습니다🙏

zangsu replied:
> 이 메서드는 이름만 set이 안들어간 setter일것 같네요!

DM 에서 해 주신 말을 고민해 본 결과, `ChessGame.end()` 라는 메서드 내에서 `state = State.END` 라고 변경될 상태를 명시해 주고 있는 것이 문제가 아닐까 하고 생각했어요. 사실, 이 부분은 `chessGame.setState(State.END)` 와 다를 바가 없으니까요.

그래서 "**상태패턴**"을 도입해 보았어요. 각 상태의 구현 클래스마다 `end()` 라는 메서드를 호출했을 때 반환되는 값을 달리 하도록이요. 이제 `InitialGame` 과 `PlayingGame`은 `end()` 메서드에서 `EndGame` 객체를 반환해 주어요. `EndGame` 은 이미 게임이 종료되었다는 예외를 터뜨리고요.

그런데, 이렇게 상태패턴을 도입한다면 setter 를 사용하는 대신 객체에 메시지를 보낸다는 것은 해결되지만, `Controller` 는 가변 객체로 남을 수 밖에 없는데요. 

1. 프로그램 자체는 상태가 변화하며 동작하는 것이기에 어딘가는 가변적이어야 하며
2. 적어도 도메인에선 최대한 불변성을 유지해 줄 수 있기 때문에

괜찮지 않을까 라고 생각하며 리뷰 요청에 포함했습니다!

choijy1705 commented on src/main/java/chess/domain/location/Location.java:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+
+    public int calculateVerticalDistance(Location target) {
+        return this.row.calculateDistance(target.row);
+    }
+
+    public int calculateHorizontalDistance(Location target) {
+        return this.column.calculateDistance(target.column);
+    }
```
rowDistance 혹은 rowDiff 로 표현하는게 어떨까요?
(columnDistance or columnDiff)
용어의 통일성도 코드를 유지 보수하는데 중요한 점일것 같아요!

zangsu replied:
프로젝트 전반적으로 `vertical`, `horizontal`을 `row`, `column`으로 변경하였습니다.

choijy1705 commented on src/main/java/chess/domain/location/Row.java:
```diff
@@ -0,0 +1,66 @@
+        for (Row row : values()) {
+            ROWS.put(row.rank, row);
+        }
+    }
```
위에 코드 참고해서 좀 더 심플하게 변경해주세요!

choijy1705 commented on src/main/java/chess/domain/location/Row.java:
```diff
@@ -0,0 +1,66 @@
+        this.rank = rank;
+    }
+
+    public static Row findByRank(String rank) {
+        try {
+            return findByRank(Integer.parseInt(rank));
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("잘못된 Row 입력입니다. 숫자를 입력해 주세요");
+        }
+    }
```
findBy는 뭔가 DB에서 조회해오는 느낌이네요!
of 혹은 from을 사용하는게 어떨까요?

zangsu replied:
`Column`에서는 정적 팩토리 메서드가 이름을 사용하는 메서드와 인덱스를 사용하는 메서드로 분리되어 있습니다. 때문에 어떤 값을 기준으로 `Column`을 생성하는지 명시해 주기 위해 `of`, `from` 이외에 다른 네이밍을 해 주게 되었어요.
`Row`는 `Column`과 정말 유사한 클래스인만큼 정적 팩토리의 형식도 일치시켜주고 싶었습니다. 

어쩌다 보니 DB 조회에서 많이 사용하는 네이밍을 사용했네요. `find`라는 동사가 문제인 것 같아 `create`로 변경해 보았습니다. 정적 팩토리 메서드이기도 하니 객체 생성의 느낌을 조금 더 담아서...

choijy1705 commented on src/main/java/chess/view/OutputView.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    private void printBoardRow(Row row, Map<Location, Piece> board) {
+        Arrays.stream(Column.values())
+                .map(column -> new Location(column, row))
+                .map(location -> board.get(location))
+                .map(this::convertPieceToString)
+                .forEach(System.out::print);
```
```suggestion
        Arrays.stream(Column.values())
            .forEach(column -> {
                Location location = new Location(column, row);
                Piece piece = board.get(location);
                System.out.println(convertPieceToString(piece));
            });
```
map 이 여러번 반복되면 코드를 읽는사람에게 피로감을 주는것 같아요!

zangsu replied:
미처 생각하지 못했던 부분인데, 다시 코드를 보니 정말 변경 과정을 따라가면서 인지적 부담이 늘어나네요. 
해당 부분은 `Row`와 다르게 역순 정렬이 필요 없기에 `for`문으로 리팩토링을 하였습니다.

choijy1705 commented on src/main/java/chess/domain/board/Path.java:
```diff
@@ -0,0 +1,97 @@
+                .allMatch(Step::isEmpty);
+    }
+
+    public boolean isEnemyAtTarget() {
+        return steps.get(findTargetIndex()).hasEnemy();
+    }
+
+    public boolean isNotAllyAtTarget() {
+        Step lastStep = steps.get(findTargetIndex());
+        return lastStep.isEmpty() || lastStep.hasEnemy();
+    }
```
같은 메서드인것 같네요🤔

zangsu replied:
`isEnemyAtTarget`은 공격을 할 수 있는 상황을 판단하기 위해 꼭 적이 있어야 합니다. 
(폰의 대각선 이동에 사용)

반면 `isNotAllyAtTarget`은 해당 위치가 아군만 아니라면 `true`를 반환합니다. 
(나머지 기물들의 이동에 사용)

---
메서드 명이 여전히 모호한가요...? 더 고민을 해보도록 하겠습니다..ㅠ

choijy1705 replied:
음...저도 딱히 떠오르는 건 없어서 괜찮을 것 같아요

choijy1705 commented on src/main/java/chess/domain/piece/implement/BlackPawn.java:
```diff
@@ -0,0 +1,15 @@
+    }
+
+    @Override
+    boolean isBackward(Path path) {
```
```suggestion
    protected boolean isBackward(Path path) {
```
아니면 접근제어자가 default인 이유가 있나요

zangsu replied:
인텔리제이에서 자동완성된 메서드를 사용하면서 고민하지 않고 놓쳤던 부분입니다...!
부모 클래스에서는 접근을 하되, 다른 클래스에서는 사용하지 못하도록 `protected` 붙여주겠습니다!

choijy1705 reviewed (changes requested):
짱수 빠르게 반영해주셨네요👍
간단한 추가 피드백 드렸어요
확인해보시고 반영부탁드려요!

zangsu committed:
refactor : 생성자에서 필드를 초기화 하도록 변경

zangsu committed:
refactor : 메서드 명에서 시도의 느낌 제거

- `tryMove()` -> `move()`

zangsu committed:
refactor : 용어 통일을 위해 vertical, horizontal 제거

zangsu committed:
refactor : 가독성을 위해 `static` 코드블럭을 `Collectors.toMap()`으로 변경

zangsu committed:
refactor : `Pawn.isBackward()`의 접근제어자 수정

zangsu committed:
refactor : 가독성을 위해 스트림을 반복문으로 변경

zangsu committed:
refactor : DB 조회와 혼동을 제거하기 위해 메서드명 변경

zangsu committed:
refactor : 게임 객체에 상태 패턴을 적용

zangsu committed:
refactor : 다양한 예외 상황에서 사용자 친화적인 메시지를 출력하도록 변경

zangsu committed:
feat : 진행중인 게임에서도 재시작을 할 수 있도록 변경

zangsu committed:
refactor : 재시작 여부를 입력할 때 y/n 의 입력이 아니라면 예외를 터뜨리도록 변경

choijy1705 commented on src/main/java/chess/view/InputView.java:
```diff
@@ -0,0 +1,33 @@
+        return SCANNER.next();
+    }
+
+    public boolean checkRestartGame() {
+        System.out.println("정말 게임을 재시작 하겠습니까? (y/n)");
+        String input = SCANNER.next();
+        if ("y".equalsIgnoreCase(input)){
+            return true;
+        }
+        if ("n".equalsIgnoreCase(input)) {
+            return false;
+        }
+        throw new IllegalArgumentException(
+                String.format("잘못된 입력입니다. y 또는 n을 입력해 주세요 (현재 입력 : %s)", input)
+        );
+    }
```
아 제가 잘못 생각했던것 같아요😭
이미 시작된 게임에 start를 하면 에러가 나는게 그렇게 이상하지는 않을텐데...
죄송합니다ㅠ
지금 방법도 좋은 방법이 될 수는 있겠네요!

choijy1705 commented on src/main/java/chess/controller/GameController.java:
```diff
@@ -0,0 +1,73 @@
+    private static final OutputView OUTPUT_VIEW = new OutputView();
+
+    private final Map<Command, Runnable> commandFunctions;
+    private ChessGame chessGame;
```
얘는 계속 바뀌어서
멤버변수로 관리되기보다는 
지역변수에서 생성하는게 좀 더 자연스러울 것 같기도 하네요!

zangsu replied:
아하, 정말 그러네요!
저도 모르게 웹의 Service 계층처럼 사용하려고 했나봐요
바로 반영해두겠습니다!

choijy1705 commented on src/main/java/chess/domain/board/Board.java:
```diff
@@ -0,0 +1,132 @@
+    public void move(Location source, Location target) {
+        Piece selectedPiece = findPieceAt(source);
+        Path path = createPath(source, target);
+        if (selectedPiece.canMove(path)) {
+            selectedPiece.move();
+            updateLocation(source, target, selectedPiece);
+            return;
+        }
+        throw new IllegalArgumentException("유효하지 않은 움직임입니다.");
```
```suggestion
        if (!selectedPiece.canMove(path)) {
            throw new IllegalArgumentException("유효하지 않은 움직임입니다.");
        }
        
        selectedPiece.move();
        updateLocation(source, target, selectedPiece);
```
반환값이 없는 메서드에서는 이런 방향이 더 안정적인 코드일것 같아요. 
누군가 리팩토링하다가 return을 빼먹거나 throw 를 지워버릴 가능성이 있을것 같아서요

zangsu replied:
뭔가 찜찜한 기분이 들었던 게 이거였나봐요!
해당 방향으로 리팩토링 해보겠습니다!

choijy1705 reviewed (approved):
짱수 피드백 반영하시느라 고생많으셨어요!
다음단계로 넘어가도 될것 같습니다👍
다음 단계도 화이팅입니다!
